L1Trigger/TrackerTFP/src/DataFormats.cc

```diff
@@ -0,0 +1,817 @@
+  Stub<Ts...>::Stub(const TTStubRef& ttStubRef, const DataFormats* dataFormats, Process p, Ts... data) :
+    dataFormats_(dataFormats),
+    p_(p),
+    frame_(ttStubRef, TTDTC::BV()),
+    data_(data...),
+    trackId_(0)
+  {
+  }
+
+  // construct StubPP from Frame
+  StubPP::StubPP(const TTDTC::Frame& frame, const DataFormats* formats) :
+    Stub(frame, formats, Process::pp)
+  {
+    for(int sectorEta = sectorEtaMin(); sectorEta <= sectorEtaMax(); sectorEta++)
+      for(int sectorPhi = 0; sectorPhi < width(Variable::sectorsPhi); sectorPhi++)
+        sectors_[sectorEta * width(Variable::sectorsPhi) + sectorPhi] = sectorsPhi()[sectorPhi];
+  }
+
+  // construct StubGP from Frame
+  StubGP::StubGP(const TTDTC::Frame& frame, const DataFormats* formats, int sectorPhi, int sectorEta) :
+    Stub(frame, formats, Process::gp), sectorPhi_(sectorPhi), sectorEta_(sectorEta)
+  {
+    const Setup* setup = dataFormats_->setup();
+    inv2RBins_ = TTBV(0, setup->htNumBinsInv2R());
+    for (int inv2R = inv2RMin(); inv2R <= inv2RMax(); inv2R++)
+      inv2RBins_.set(inv2R + inv2RBins_.size() / 2);
+  }
+
+  // construct StubGP from StubPP
+  StubGP::StubGP(const StubPP& stub, int sectorPhi, int sectorEta) :
+    Stub(stub, stub.r(), stub.phi(), stub.z(), stub.layer(), stub.inv2RMin(), stub.inv2RMax()),
+    sectorPhi_(sectorPhi),
+    sectorEta_(sectorEta)
+  {
+    const Setup* setup = dataFormats_->setup();
+    get<1>(data_) -= (sectorPhi_ - .5) * setup->baseSector();
+    get<2>(data_) -= (r() + dataFormats_->chosenRofPhi()) * setup->sectorCot(sectorEta_);
+    dataFormats_->convertStub(data_, frame_.second, p_);
+  }
+
+  // construct StubHT from Frame
+  StubHT::StubHT(const TTDTC::Frame& frame, const DataFormats* formats, int inv2R) :
+    Stub(frame, formats, Process::ht), inv2R_(inv2R)
+  {
+    fillTrackId();
+  }
+
+  // construct StubHT from StubGP
+  StubHT::StubHT(const StubGP& stub, int phiT, int inv2R) :
+    Stub(stub, stub.r(), stub.phi(), stub.z(), stub.layer(), stub.sectorPhi(), stub.sectorEta(), phiT),
+    inv2R_(inv2R)
+  {
+    get<1>(data_) -= format(Variable::inv2R).floating(this->inv2R()) * r() + format(Variable::phiT).floating(this->phiT());
+    fillTrackId();
+    dataFormats_->convertStub(data_, frame_.second, p_);
+  }
+
+  void StubHT::fillTrackId() {
+    TTBV ttBV(bv());
+    trackId_ = ttBV.extract(width(Variable::sectorPhi) + width(Variable::sectorEta) + width(Variable::phiT));
+  }
+
+  // construct StubMHT from Frame
+  StubMHT::StubMHT(const TTDTC::Frame& frame, const DataFormats* formats) :
+    Stub(frame, formats, Process::mht)
+  {
+    fillTrackId();
+  }
+
+  // construct StubMHT from StubHT
+  StubMHT::StubMHT(const StubHT& stub, int phiT, int inv2R) :
+    Stub(stub, stub.r(), stub.phi(), stub.z(), stub.layer(), stub.sectorPhi(), stub.sectorEta(), stub.phiT(), stub.inv2R())
+  {
+    const Setup* setup = dataFormats_->setup();
+    get<6>(data_) = this->phiT() * setup->mhtNumBinsPhiT() + phiT;
```

Comment:
Add comment, "update track (phIT, inv2R), and phi residuals w.r.t. track, to reflect MHT cell assignment."

Reply:
done